plugins/kubernetes/app/assets/javascripts/kubernetes/controllers/kubernetes_dashboard_ctrl.js

```diff
@@ -1,15 +1,184 @@
+        )
+        .then(function() {
+            startListeningForUpdates()
+          }
+        );
+    }
+
+    init();
+
+    /*
+     Controller private functions.
+     */
+
+    function loadEnvironments() {
+      $scope.loadingEnvironments = true;
+
+      var deferred = $q.defer();
+      environmentsService.loadEnvironments($scope.project_id).then(
+        function(environments) {
+          $scope.loadingEnvironments = false;
+          $scope.environments = environments;
+
+          // Selects the production environment by default
+          var default_environment = _.findWhere(environments, {'is_production': true});
+          $scope.environment = _.isDefined(default_environment) ? default_environment : environments[0];
+
+          deferred.resolve();
+        },
+        function(result) {
+          $scope.loadingEnvironments = false;
+          result.messages.map(function(message) {
+            notificationService.error(message);
+          });
+          deferred.reject();
+        }
+      );
+
+      return deferred.promise;
+    }
+
+    function loadClusterState() {
+      $scope.loadingClusterState = true;
+      $scope.dashboard_data = [];
+
+      var deferred = $q.defer();
+      kubernetesService.loadDashboardData($scope.project_id, $scope.environment).then(
+        function(data) {
+          $scope.loadingClusterState = false;
+          if (_.isNotUndefinedOrEmpty(data)) {
+            $scope.dashboard_data = data;
+            deferred.resolve();
+          }
+          else {
+            deferred.reject();
+          }
+        },
+        function(result) {
+          $scope.loadingClusterState = false;
+          result.messages.map(function(message) {
+            notificationService.error(message);
+          });
+          deferred.reject();
+        }
+      );
+
+      return deferred.promise;
+    }
+
+    function startListeningForUpdates() {
+      SseFactory.on('k8s', function(msg) {
+        var role = findOrCreateRole(msg);
+        var deploy_group = findOrCreateDeployGroup(role, msg);
+        var release = findOrCreateRelease(deploy_group, msg);
+        updateReleaseState(release, msg);
+      });
+    }
+
+    function findOrCreateRole(msg) {
+      var role = findRole(msg.role);
+
+      if (_.isUndefined(role)) {
+        role = {
+          name: msg.role,
+          deploy_groups: []
+        };
+        $scope.dashboard_data.push(role);
+      }
+
+      return role;
+    }
+
+    function findOrCreateDeployGroup(role, msg) {
+      if (_.isDefined(role)) {
+        var deploy_group = findDeployGroup(role, msg.deploy_group);
+
+        if (_.isUndefined(deploy_group)) {
+          deploy_group = {
+            name: msg.deploy_group,
+            releases: []
+          };
+          role.deploy_groups.push(deploy_group);
+        }
+
+        return deploy_group;
+      }
+    }
+
+    function findOrCreateRelease(deploy_group, msg) {
+      if (_.isDefined(deploy_group)) {
```

Comment:
Probably not needed since you're creating the deploy_group if it doesn't exist. Plus it'll throw an exception anyway if the deploy_group is nil later @ https://github.com/zendesk/samson/pull/685/files#diff-1fe6543a8989338e21335956e8dd9daeR163 when it tries to deference an undefined release.